pytorch lightning docs improvements

kamil-kaczmarek opened the pull request:
PyTorch Lightning integration docs.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
-Requirements
-------------
-This integration is enabled as part of PyTorch Ligthning loggers module, so all you need is to have |neptune-client| installed.
+* preview running experiment
```
Not 100% that preview is correct here. 
Monitor live or see experiment as it is running is clear and simple. 

I am good with preview if we have users using this term -> just don't have that info yet.

kamil-kaczmarek replied:
I'll use "see experiment as it is running"

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
+* save model checkpoints.
 
-.. code-block:: bash
+.. tip::
```
I think we should have 3 things close to the top (perhpas the tip or note are great options).

- link to project in Neptune
- link to full code that runs it (so that you know there is some code that is tested)
- link to colab if applicable (just so you know you can) same as full code

kamil-kaczmarek replied:
corrected.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
 
-Log additional information after the **.fit** loop ends
--------------------------------------------------------
+**Install PyTorch Lightning and torchvision**
```
This should be in pre-requisites with a link to their documentation on how to install I believe -> it is not Neptune related.

kamil-kaczmarek replied:
corrected.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
-Nothing changes here, just your typical ``LightningModule`` structure.
+Introduction
+^^^^^^^^^^^^
+This quickstart will show you how to use Neptune with your existing PyTorch Lightning code. You have three options:
```
I think this is not true. 
The quickstart doesn't show you how to connect -> it shows you how to log some experiment to lightning. 

We could have how-to connect as well -> perhaps as a third piece of content but for now I'd rather not have it :)

kamil-kaczmarek replied:
thanks, corrected.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
-        @pl.data_loader
-        def train_dataloader(self):
-            ...
+* ``pytorch-lightning>=0.9.0``
```
I think we should have supported versions somewhere at the top, just below the ::tip:: perhaps?

kamil-kaczmarek replied:
thx, corrected.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
-**Log additional metrics**
+Step 1 - Necessary Imports
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+No a big deal here, just imports. Note ``pytorch_lightning`` at the bottom.
```
I'd go with "Your typical PyTorch imports. Note..."

kamil-kaczmarek replied:
ok.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
+Now, you have all dependencies installed. Let's move to the actual integration.
 
-**Log additional metrics**
+Step 1 - Necessary Imports
```
I think it should be an action so 

Step 1 - import PyTorch libs and lightning

Also I think we are using Step 1: in other places (could you check quickstarts? )

kamil-kaczmarek replied:
I will follow `Step 1:` - thanks

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
+    import pytorch_lightning as pl
 
-**Log performance charts**
+Step 2 - Model Hyper-Parameters
```
I'd change it to action -> Log model hyper-parameters

kamil-kaczmarek replied:
corrected.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
-**Log performance charts**
+Step 2 - Model Hyper-Parameters
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+You will see them in Neptune parameters tab.
```
I think it is better (read it somewhere and github goes with this actually).

What to do

Code

Explain what it does

kamil-kaczmarek replied:
nice idea! LIke it, correcting.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
+              'batch_size': 32}
 
-    neptune_logger.experiment.log_image('confusion_matrix', fig)
+Step 3 - Define Lightning Module
```
Here you go with Define, I like it!

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
-    neptune_logger.experiment.log_image('confusion_matrix', fig)
+Step 3 - Define Lightning Module
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+This is minimal example of the ``pl.LightningModule``. Notice the Cross Entropy loss in the ``training_step`` method.
```
Hmm, how is cross entropy important for me learning how to log things?

kamil-kaczmarek replied:
It is the only metric that will be logged to Neptune. I added info that this will be logged to Neptune, so that user see where exactly logging happens.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
+This is minimal example of the ``pl.LightningModule``. Notice the Cross Entropy loss in the ``training_step`` method.
 
-**Save checkpoints folder after training**
+Also, note that you pass learning rate from the ``PARAMS`` dictionary.
```
This perhaps as Note ?

kamil-kaczmarek replied:
fixed.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
+            return torch.optim.Adam(self.parameters(), lr=PARAMS['LR'])
 
-Closing is optional but you may want to close it and then do something afterwards.
+Step 4 - Prepare Data Loader
```
Those two steps feel like one step to me in this context. 

We are teaching people how to do Neputne not lightning so I'd put Loader and System in one Step for sure.

Add two substeps of, course. 
1. 
2.

kamil-kaczmarek replied:
agreed.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
-Now you can watch your pytorch-lightning model training in Neptune!
+Step 5 - Create NeptuneLogger
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+``NeptuneLogger`` is an object that integrates Neptune with PyTorch Lightning allowing you to track your experiments in Neptune. It's a part of the lightning library.
```
I'd put those after the code with all the notes. 
first show, then explain -> it is easier to understand (and skip if I don't care :) ).

kamil-kaczmarek replied:
Fro now on I follow what do to / code / explain scheme.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
+
+.. tip::
+
+    Make sure to use your API token in your projects. Read more about how to |token|.
```
this link shouldn't be in the external links but rather with this :ref:`get and set token <XXX>`.
In the latest docs version all those links that are internal are fixed.

kamil-kaczmarek replied:
I'm not sure here. ref will not open new browser tab which is preferred behavior - when clicking I will not lose the context.

jakubczakon replied:
If we don't use :ref: we run the risk (and that was the case for the past few weeks) of having links that move you to the home of the docs (beacuse they are wrong). 

I think it should be :ref: there just maybe is an option to open in new tab. 
Could you look into it?

kamil-kaczmarek replied:
thanks, corrected.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
+
+    Make sure to use your API token in your projects. Read more about how to |token|.
+
+Step 6 - Fit model to the data - Neptune tracks it automatically
```
The step should be pass logger to the trainer I think.

kamil-kaczmarek replied:
I chanegd this step a bit.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
-    import os
+    trainer = pl.Trainer(max_epochs=PARAMS['max_epochs'],
+                         logger=neptune_logger)
+    model = LitModel()
```
... then this would be an additional step -> run experiments

kamil-kaczmarek replied:
agreed.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
-            self.l1 = torch.nn.Linear(28 * 28, 10)
+.. image:: ../_static/images/integrations/pytorchlightning_neptuneml.png
+   :target: ../_static/images/integrations/pytorchlightning_neptuneml.png
+   :alt: PyTorchLightning neptune.ai integration
```
some link to experiment shown in image would be good I think below (I know there are some up but you know how it is... :) )

jakubczakon replied:
I'd also add Colab button somewhere here -> in case someone feels actionable enough to try it.

kamil-kaczmarek replied:
added colab and link.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
+#. |skorch|
 
-    from pytorch_lightning.loggers.neptune import NeptuneLogger
+You may also like these two integrations:
```
I don't think those are related -> I'd rather show optuna or something

kamil-kaczmarek replied:
I see. I put optuna and plotly.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,305 @@
-        def test_dataloader(self):
-            # OPTIONAL
-            return DataLoader(MNIST(os.getcwd(), train=False, download=True, transform=transforms.ToTensor()), batch_size=BATCHSIZE)
+Other integrations you may like
```
I like this section :)

kamil-kaczmarek replied:
thanks!

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,315 @@
-   :target: ../_static/images/integrations/pytorchlightning_neptuneml.png
+What will you get?
+------------------
+.. image:: ../_static/images/integrations/lightning_basic.gif
```
is this .gif available? I don't see it in Files changed for this PR

jakubczakon replied:
Also, perhaps it would be better to just create a YT and embed -> I think this is a longer "piece" and having a .gif here is not the best option (as we discussed before)

kamil-kaczmarek replied:
ups, sorry did not put them here. For now it's gif, I'll do video at the end.

kamil-kaczmarek replied:
Then YT and embed of course.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,315 @@
+
+.. note::
+
+    This integration is tested with ``pytorch-lightning==0.9.0``. See 'Troubleshoot' section for information about other versions.
```
and neptun-client XXX right?

jakubczakon replied:
please add link to `Troubleshoot` via :ref:XXX would be easiest I presume

kamil-kaczmarek replied:
corrected.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,315 @@
+----------
+This quickstart will show you how to log PyTorch Lightning experiments to Neptune using ``NeptuneLogger`` (part of the pytorch-lightning library).
+
+Check this public project with example experiments: |project|.
```
I must have miscoummunicated but I think:
- project
- script-basic (maybe script-advanced)
- colab-basic (maybe colab-advanced)

Should be a part of the what will you get or we can have a section with some other name above quickstart and below what will you get. 

Definitely |project| shouldn't be the first thing here. It doesn't help the person go through quickstart does it?
It can show what you get nicely -> hence I feel it makes sense to have it above in what you will get. 

When it comes to script and colab the reason for putting it up there (not sure if my "implementation" is good) is to show people right of the bat that there is a code that will work that they can use (come back to)

kamil-kaczmarek replied:
ok, cool, corrected.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,315 @@
+* Python 3,
+* Neptune account |register|,
+* ``pytorch-lightning==0.9.0`` and ``torchvision`` installed (lightning |lightning-install|),
+* Minimal familiarity with the PyTorch Lightning.
```
link to some demo about lightnign or their readme could be helpful.

kamil-kaczmarek replied:
added.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,315 @@
+* ``pytorch-lightning==0.9.0`` and ``torchvision`` installed (lightning |lightning-install|),
+* Minimal familiarity with the PyTorch Lightning.
+
+Installation
```
I think I'd change it into Step 0: Install neptune or something. 

...or maybe this should be in the prerequisites. 
Instead of register have this -> with link to installation parts of the docs. 
Remember that people don't have to be registered to use/try it.

kamil-kaczmarek replied:
I moved all to the prerequisited subsection.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,315 @@
+              'learning_rate': 0.005,
+              'batch_size': 32}
+
+Hyper-parameters are passed in the regular Python dictionary. You will see them logged in Neptune parameters tab.
```
are passed to what -> in this section the reader doesn't yet know how the logger looks or that it exists really. 
This should be more explicit with less assumptions about what reader knows I feel.

kamil-kaczmarek replied:
Agree, that User does not know much about logger at this point. I wanted co communicate that you just have prepared params in python dict -> code above is just this.
I need to change wording or edit these two sentences and it will be clear :)

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,315 @@
+
+Step 3: Define LightningModule and DataLoader
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Implement minimal example of the ``pl.LightningModule`` and simple ``DataLoader``.
```
I like it, it is concise and focused now.

kamil-kaczmarek replied:
thanks!

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,315 @@
-        @pl.data_loader
-        def train_dataloader(self):
-            ...
+Cross Entropy loss in the ``training_step`` method will be logged to Neptune in every training step. ``DataLoader`` is a purely PyTorch object. Notice, that you pass ``learning_rate`` and ``batch_size`` from the ``PARAMS`` dictionary - all params will be logged as experiment parameters.
```
perhaps we could add "...as 'train_loss'" and will be visible in the Logs section.

jakubczakon replied:
I think it should read "is a pure PyTorc object"

kamil-kaczmarek replied:
corrected.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,315 @@
-        @pl.data_loader
-        def train_dataloader(self):
-            ...
+Cross Entropy loss in the ``training_step`` method will be logged to Neptune in every training step. ``DataLoader`` is a purely PyTorch object. Notice, that you pass ``learning_rate`` and ``batch_size`` from the ``PARAMS`` dictionary - all params will be logged as experiment parameters.
```
I think extracting this line here and presenting it could be useful to focus the reader. 

```python
result.log('train_loss', loss)
```

kamil-kaczmarek replied:
added.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,315 @@
 
-Log additional information after the **.fit** loop ends
--------------------------------------------------------
+    Make sure to use your API token in your projects. Read more about how to |token|.
```
I don't know if it should be make sure or something like if you want to use in your project. 

Some people may just want to run the how-to and there is no must there

kamil-kaczmarek replied:
corrected.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,315 @@
+Check this experiment |exp-link|.
 
-    model_checkpoint = pl.callbacks.ModelCheckpoint(filepath='my/checkpoints')
+|Run on Colab|
```
We should have a CTA see the full code or something as well as colab.
Some people may just want to copy the code.

kamil-kaczmarek replied:
added.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,315 @@
-    trainer = Trainer(logger=neptune_logger,
-                      checkpoint_callback=model_checkpoint)
-    trainer.fit(model)
+Advanced options
```
The Advanced options are on a TODO list I presume.

kamil-kaczmarek replied:
yes

kamil-kaczmarek commented:
Hey @jakubczakon thanks for all the comments/suggestions!

I applied a lot of small changes, now I continue with "advanced" section. I'll ping you again once done.

thanks again!

kamil-kaczmarek commented:
Hey @jakubczakon - docs for Pytorch Lightning are ready for review.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,744 @@
-   :target: ../_static/images/integrations/pytorchlightning_neptuneml.png
+What will you get?
+------------------
+.. image:: ../_static/images/integrations/lightning_basic.gif
```
Could you change this to a video -> I think with video you could show/talk more about what you get and what you need to add to do it.  Also, it would be more reusable. 

I suggest:
- 30-60s
- it could start in the code -> then run from terminal -> go to the app (or skip the terminal)
- add an inviting thumbnail -> if you use this loom thing I think we can get a really cool one with your face even!!

kamil-kaczmarek replied:
Agree 101% :) I'll do it, and let you know.

kamil-kaczmarek replied:
Added video

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,744 @@
+
+Notice ``pytorch_lightning`` at the bottom.
+
+Step 2: Define Hyper-Parameters
```
I started wondering. Should this be the basic usage or advanced use already. 
What do you think?

kamil-kaczmarek replied:
We talked about it :)

jakubczakon commented:
https://www.loom.com/share/4a14b72b61534bc8b212bff08d2bc1bc

kamil-kaczmarek commented:
Hey @jakubczakon,

Thanks for the feedback - especially about the structure of "advanced options" and images that show effects of the logging actions.

New version is ready for review.

Changed structure of "advance options", added images, corrected links, ported to 1.x. Also, changes some details in the quickstart.

jakubczakon commented:
@kamil-kaczmarek let's wait for the neptune-examples to be merged first. 

Once we have that ready, we should update links here (make sure they are working) and we can merge it.

kamil-kaczmarek commented:
@jakubczakon 
added video
all Colab links checked.

kamil-kaczmarek commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,744 @@
+
+Where to start?
+---------------
+To get started with this integration, follow whatever is most convenient to you:
```
done.

kamil-kaczmarek commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,744 @@
-Monitor your PyTorch Lightning training in Neptune
---------------------------------------------------
-Now you can watch your pytorch-lightning model training in Neptune!
+#. Follow the walk-through below for detailed explanations,
```
This refers to the old commit, it no longer exist in the newer ones. I made sure that colab/code are described as supplementary material.

kamil-kaczmarek commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,744 @@
+----
 
-    neptune_logger.experiment.stop()
+Advanced options
```
It's already there, take a look at the section: "Jump to your favorite option".

kamil-kaczmarek replied:
Overall idea is that you have:

"Advanced options" section with intro + links to colab/code/project, then, you go to the "Before you start" to see what you need to install to start working with advance options, then, in the "Jump to your favorite option" you have listing of all options.

Is that ok? What do you think @jakubczakon

kamil-kaczmarek commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,744 @@
+
+.. _quickstart:
+
+Quickstart
```
added.

kamil-kaczmarek commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,744 @@
+
+Step 3: Define LightningModule, LightningDataModule and Callbacks
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Step 3.1: Implement LightningModule
```
Fixed.

kamil-kaczmarek commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,719 @@
-The integration enables you to log |PyTorch Lightning| metrics to Neptune.
+What will you get?
+------------------
+[VIDEO PLACEHOLDER]
```
Please take a look at the latest commit. There is video added.

kamil-kaczmarek commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,723 @@
+You have ``Python 3.x`` and following libraries installed:
+
+* ``neptune-client==0.4.123`` or newer: See :ref:`neptune-client installation guide <installation-neptune-client>`.
+* ``pytorch`` and ``torchvision``. See |pytorch-install|.
```
fixed.

kamil-kaczmarek commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,723 @@
+
+#. :ref:`Quickstart <quickstart>` below for detailed explanations,
+#. Open Colab notebook (badge-link below) with quickstart code and run it as a "`neptuner`" user - zero setup, it just works,
+#. View quickstart code as a plain Python script on |script|.
```
thanks! Fixed.

kamil-kaczmarek commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,723 @@
 
-    model = CoolSystem()
-    trainer = Trainer(max_epochs=10, logger=neptune_logger)
+Check this experiment |exp-link| or view quickstart code as a plain Python script on |script|.
```
thanks. Fixed.

kamil-kaczmarek commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,723 @@
+---------------
+To get started with this integration, follow whatever is most convenient to you:
+
+#. :ref:`Quickstart <quickstart>` below for detailed explanations,
```
sure, corrected.

kamil-kaczmarek commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,744 @@
+
+Here, you log test loss.
+
+Step 7: Run additional actions
```
I'm afraid that you're commenting and old commit. Now, images are added to each advanced option.

kamil-kaczmarek commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,719 @@
-The integration enables you to log |PyTorch Lightning| metrics to Neptune.
+What will you get?
+------------------
+[VIDEO PLACEHOLDER]
```
Video is now available.

kamil-kaczmarek commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,744 @@
+--------------------
+The fastest way is to simply chat with us. Chat icon is located directly in-app, in the lower right corner. Use it!
+
+.. image:: ../_static/images/integrations/chat-icon.png
```
Sure, changed.

kamil-kaczmarek commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,744 @@
+---------------
+This integration is tested with ``pytorch-lightning==0.9.0`` and ``neptune-client==0.4.122``. Make sure that you use correct versions.
+
+How to ask for help?
```
yep, corrected. :)

kamil-kaczmarek commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,744 @@
-    # You can stop the experiment
     neptune_logger.experiment.stop()
 
+In the :ref:`Step 4 <adv-step-4>` we created ``NeptuneLogger`` with ``close_after_fit=False``, so we need to close Neptune experiment explicitly at the end.
```
fixed.

kamil-kaczmarek commented:
Hey @jakubczakon,

Thanks for the review. I applied all corrections.

jakubczakon commented on docs/integrations/pytorch_lightning.rst:
```diff
@@ -1,318 +1,727 @@
+
+.. tip::
+
+    Are you a Pro User? Take a look at the :ref:`advanced options <advanced-options>`.
```
Please change the Pro User part to something else as it sounds as if it could be a Neptune paid plan

kamil-kaczmarek replied:
fixed.
